### ISSUE TITLE
Add .exists() and .iscontract() on to the address type

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#4910]
Not fully sure yet if `.exists()` can be implemented properly, but `.iscontract()` should be possible to be implemented via the `extcodesize` instruction.

With having `.iscontract()`, #4823 would not need to be implemented.

**Comment by frangio** [comment]
What would be the semantics of `.exists()`?

**Comment by axic** [comment]
`.exists()` probably should be postponed until there is a consensus on what that means in Ethereum, however when looking at the account cleanup rules during creation/selfdestruct/etc. should be a good indication.

I actually don't think this can be done properly without support form a new EVM opcode, because one could check for balance and code size, but cannot check for storage, which is possible to be non-empty if the account is only used through delegatecalls.

**Comment by wadealexc** [comment]
I would be cautious of any implementation of `.iscontract()` -- the whole idea of testing `iscontract` has led to lots of contracts vulnerable to exploits done by constructors

**Comment by axic** [comment]
Actually it seems that `.iscontract()` could be implemented by just using `extcodehash`.

**Comment by axic** [comment]
`.exists()` can be clearly implemented post Constantinople, because [EIP-1052](https://eips.ethereum.org/EIPS/eip-1052) for `EXTCODEHASH` clearly defines it:
> In case the account does not exist or is empty (as defined by EIP-161) 0 is pushed to the stack.

From [EIP-161](https://eips.ethereum.org/EIPS/eip-161):
> An account is considered empty when it has no code and zero nonce and zero balance.

Similarly `.iscontract()` is clearly defined with EIP-1052:
> In case the account does not have code the keccak256 hash of empty data (i.e. c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470) is pushed to the stack.

**Comment by axic** [comment]
@frangio @wadeAlexC any change of view since these are provided by opcodes now?

**Comment by wadealexc** [comment]
Since `EXTCODEHASH` returns a nonzero value during the constructor, this seems fine on the surface.

**Comment by frangio** [comment]
I don't think that really mitigates the vulnerabilities that can arise from using `.isContract()`, because an account can pass the `EXTCODEHASH` test but still become a contract in the future.

We do have an `isContract` function in OpenZeppelin though, because some ERCs require using it. So I think I would be okay with including it in Solidity. But it needs to be used carefully.

**Comment by chriseth** [comment]
Would calling it `hasCode` help in any way?

**Comment by axic** [comment]
> I don't think that really mitigates the vulnerabilities that can arise from using .isContract(), because an account can pass the EXTCODEHASH test but still become a contract in the future.

Are we making commitments in the language that a property of an account in the state cannot ever change? :wink:

@frangio Could you elaborate this sentence?

**Comment by axic** [comment]
This would become very useful in conjunction with https://github.com/ethereum/EIPs/pull/8838

On non-EOF targets it could fallback to extcodesize/extcodehash.
